core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java

```diff
@@ -15,52 +15,176 @@
+
+  private CloudSqlInstance instance;
 
   @Before
-  public void setup() {
-    MockitoAnnotations.openMocks(this);
-    when(googleCredentials.createScoped("https://www.googleapis.com/auth/sqlservice.login"))
-        .thenReturn(scopedCredentials);
+  public void setup() throws Exception {
+    MockAdminApi mockAdminApi = new MockAdminApi();
+    this.keyPairFuture = Futures.immediateFuture(mockAdminApi.getClientKeyPair());
+    executorService = newTestExecutor();
+    refreshCount = 0;
+    sslData = new SslData(null, null, null);
+    data = new InstanceData(null, sslData, null);
+  }
+
+  @After
+  public void teardown() {
+    executorService.shutdownNow();
   }
 
   @Test
-  public void downscopesGoogleCredentials() {
-    GoogleCredentials downscoped = CloudSqlInstance.getDownscopedCredentials(googleCredentials);
-    assertThat(downscoped).isEqualTo(scopedCredentials);
-    verify(googleCredentials, times(1))
-        .createScoped("https://www.googleapis.com/auth/sqlservice.login");
+  public void testCloudSqlInstanceDataRetrievedSuccessfully() throws Exception {
+    fetcher =
+        (instanceName, accessTokenSupplier, authType, executor, keyPair) -> {
+          Thread.sleep(100);
+          refreshCount++;
+          this.data =
+              new InstanceData(null, sslData, Date.from(Instant.now().plus(1, ChronoUnit.HOURS)));
+          return data;
+        };
+    // initialize instance after mocks are set up
+    instance = newCloudSqlInstance();
+
+    SslData gotSslData = instance.getSslData();
+    Truth.assertThat(gotSslData).isSameInstanceAs(this.sslData);
+    Truth.assertThat(refreshCount).isEqualTo(1);
   }
 
   @Test
-  public void throwsErrorForWrongCredentialType() {
-    RuntimeException ex =
-        assertThrows(
-            RuntimeException.class,
-            () -> CloudSqlInstance.getDownscopedCredentials(oAuth2Credentials));
-
-    assertThat(ex)
-        .hasMessageThat()
-        .contains("Failed to downscope credentials for IAM Authentication");
+  public void testInstanceFailsOnConnectionError() throws Exception {
+    fetcher =
+        (instanceName, accessTokenSupplier, authType, executor, keyPair) -> {
+          throw new ExecutionException(new IOException("Fake connection error"));
+        };
+
+    // initialize instance after mocks are set up
+    instance = newCloudSqlInstance();
+
+    RuntimeException ex = Assert.assertThrows(RuntimeException.class, instance::getSslData);
+    Truth.assertThat(ex).hasMessageThat().contains("Fake connection error");
+  }
+
+  @Test
+  public void testCloudSqlInstanceForcesRefresh() throws Exception {
+    fetcher =
+        (instanceName, accessTokenSupplier, authType, executor, keyPair) -> {
+          Thread.sleep(100);
+          refreshCount++;
+          this.data =
+              new InstanceData(null, sslData, Date.from(Instant.now().plus(1, ChronoUnit.HOURS)));
+          return data;
+        };
+
+    instance = newCloudSqlInstance();
```

Comment:
Let's inline this function -- it's a little bit of noise, but removes the indirection and helps make the test clearer.

Reply:
That also helps remove the need to rely on class level variables (i.e., fetcher).

Reply:
Done. I got rid of most of the fields, moved them to instance variables.